Azure backend: support snapshots/versioning

evenh opened the pull request:
Rebased PR with some small fixes. Originally by @pmarques ([2018](https://github.com/hashicorp/terraform/pull/18512)) and @rahdjoudj ([2019](https://github.com/hashicorp/terraform/pull/21888)).

Fixes #18284 
Closes #18512 
Closes #21888

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=24069)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=24069)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**1** out of **2** committers have signed the CLA.

- [x] rahdjoudj
- [ ] evenh

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform?pullRequest=24069).</sub>

Bessonov commented:
Just mess up my state. @tombuildsstuff please, don't let this MR rotten in 2020. If you have not enough time, who can help out?

WebSpider commented:
Really looking forward to this!
(Just spent a few hr rolling back a failed provider update, rewriting statefile by hand)

landro commented:
Hi @tombuildsstuff , any chance we can get this merged any time soon? We urgently need this in order to prevent incidents like the one @WebSpider experienced. Thanks!

WascawyWabbit commented:
Really looking forward to this, any news on when it will be released?

Bowbaq commented:
@tombuildsstuff I just messed my state, thought I was just going to restore a previous version & came to find out that versioning is not only not enabled by default but not even possible. This would have likely saved me a few hours of work

WascawyWabbit commented:
Any update on this one please? I can see that review requests are pending From Feb 10th, which is > 2 months ago now.

doubletr commented:
@tombuildsstuff, Could you please help to review this RP? It's pending for about 3 months. Thanks for your help in advance.

tombuildsstuff commented:
Dependent on #24669

evenh commented on backend/remote-state/azure/client.go:
```diff
@@ -67,6 +70,17 @@ func (c *RemoteClient) Put(data []byte) error {
 	ctx := context.TODO()
+
+	if c.versioning {
+		snapshotInput := blobs.SnapshotInput{LeaseID: options.LeaseID}
```
The original approach didn't use a specific lease ID (just `nil`). I'm not sure what the best approach is here.

tombuildsstuff replied:
my understanding of the Azure API is that a Lease ID is required if one is configured (and will error if not) - so the tests should confirm this

evenh commented:
I have squashed the commits and rebased after #24669 was merged in. Could you please review @tombuildsstuff, especially since you are very familiar with Giovanni ;)

tombuildsstuff commented on backend/remote-state/azure/backend.go:
```diff
@@ -51,6 +51,13 @@ func New() backend.Backend {
 				DefaultFunc: schema.EnvDefaultFunc("ARM_SAS_TOKEN", ""),
 			},
 
+			"versioning": {
```
whilst the S3 Backend calls this "versioning" (due to that being [the name of the feature in S3](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/enable-versioning.html)) - since Azure's calling this "Snapshots" instead could we use that terminology instead?

```suggestion
			"snapshot": {
```

tombuildsstuff commented on backend/remote-state/azure/backend.go:
```diff
@@ -51,6 +51,13 @@ func New() backend.Backend {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Enable/Disable automatic blob snapshotting",
+				DefaultFunc: schema.EnvDefaultFunc("ARM_VERSIONING", false),
```
```suggestion
				DefaultFunc: schema.EnvDefaultFunc("ARM_SNAPSHOT", false),
```

tombuildsstuff commented on backend/remote-state/azure/client.go:
```diff
@@ -67,6 +70,17 @@ func (c *RemoteClient) Put(data []byte) error {
+		snapshotInput := blobs.SnapshotInput{LeaseID: options.LeaseID}
+
+		if _, err := c.giovanniBlobClient.Snapshot(ctx, c.accountName, c.containerName, c.keyName, snapshotInput); err != nil {
+			return err
```
could we wrap this error to be a little clearer here:

```suggestion
			return fmt.Errorf("Error snapshotting Blob %q (Container %q / Account %q): %+v", c.keyName, c.containerName, c.accountName, err)
```

tombuildsstuff commented on backend/remote-state/azure/client.go:
```diff
@@ -67,6 +70,17 @@ func (c *RemoteClient) Put(data []byte) error {
+	if c.versioning {
+		snapshotInput := blobs.SnapshotInput{LeaseID: options.LeaseID}
+
+		if _, err := c.giovanniBlobClient.Snapshot(ctx, c.accountName, c.containerName, c.keyName, snapshotInput); err != nil {
```
can we add a debug log before this too:

```suggestion
		log.Printf("[DEBUG] Snapshotting existing Blob %q (Container %q / Account %q)", c.keyName, c.containerName, c.accountName)
		if _, err := c.giovanniBlobClient.Snapshot(ctx, c.accountName, c.containerName, c.keyName, snapshotInput); err != nil {
```

tombuildsstuff commented on website/docs/backends/types/azurerm.html.md:
```diff
@@ -154,6 +154,8 @@ The following configuration options are supported:
 
 * `environment` - (Optional) The Azure Environment which should be used. This can also be sourced from the `ARM_ENVIRONMENT` environment variable. Possible values are `public`, `china`, `german`, `stack` and `usgovernment`. Defaults to `public`.
 
+* `versioning` - (Optional) Should versioning for state file be used - default `false`. This value can also be sourced from the `ARM_VERSIONING` environment variable.
```
```suggestion
* `snapshot` - (Optional) Should the Blob used to store the Terraform Statefile be snapshotted before use? Defaults to `false`. This value can also be sourced from the `ARM_SNAPSHOT` environment variable.
```

tombuildsstuff commented on backend/remote-state/azure/backend.go:
```diff
@@ -150,6 +157,7 @@ type Backend struct {
 	containerName string
 	keyName       string
 	accountName   string
+	versioning    bool
```
```suggestion
	snapshot    bool
```

tombuildsstuff commented on backend/remote-state/azure/backend.go:
```diff
@@ -180,6 +188,7 @@ func (b *Backend) configure(ctx context.Context) error {
 	b.containerName = data.Get("container_name").(string)
 	b.accountName = data.Get("storage_account_name").(string)
 	b.keyName = data.Get("key").(string)
+	b.versioning = data.Get("versioning").(bool)
```
```suggestion
	b.snapshot = data.Get("snapshot").(bool)
```

tombuildsstuff commented on backend/remote-state/azure/backend_state.go:
```diff
@@ -90,6 +90,7 @@ func (b *Backend) StateMgr(name string) (state.State, error) {
 		containerName:      b.containerName,
 		keyName:            b.path(name),
 		accountName:        b.accountName,
+		versioning:         b.versioning,
```
```suggestion
		snapshot:         b.snapshot,
```

tombuildsstuff commented on backend/remote-state/azure/backend_test.go:
```diff
@@ -21,6 +21,7 @@ func TestBackendConfig(t *testing.T) {
 		"storage_account_name": "tfaccount",
 		"container_name":       "tfcontainer",
 		"key":                  "state",
+		"versioning":           false,
```
```suggestion
		"snapshot":           false,
```

tombuildsstuff commented on backend/remote-state/azure/backend_test.go:
```diff
@@ -33,6 +34,9 @@ func TestBackendConfig(t *testing.T) {
 	if b.keyName != "state" {
 		t.Fatalf("Incorrect keyName was populated")
 	}
+	if b.versioning != false {
```
```suggestion
	if b.snapshot != false {
```

tombuildsstuff commented on backend/remote-state/azure/backend_test.go:
```diff
@@ -33,6 +34,9 @@ func TestBackendConfig(t *testing.T) {
 		t.Fatalf("Incorrect keyName was populated")
 	}
+	if b.versioning != false {
+		t.Fatalf("Incorrect versioning was populated")
```
```suggestion
		t.Fatalf("Incorrect snapshot was populated")
```

tombuildsstuff commented on backend/remote-state/azure/client.go:
```diff
@@ -24,6 +26,7 @@ type RemoteClient struct {
 	containerName      string
 	keyName            string
 	leaseID            string
+	versioning         bool
```
```suggestion
	snapshot         bool
```

tombuildsstuff commented on backend/remote-state/azure/client.go:
```diff
@@ -67,6 +70,17 @@ func (c *RemoteClient) Put(data []byte) error {
 
 	ctx := context.TODO()
+
+	if c.versioning {
```
```suggestion
	if c.snapshot {
```

landro commented:
Hi there, just wanted to let you know that if you enable soft delete on the storage account, you actually get new snapshots of the state every time you run terraform. See this for details: https://docs.microsoft.com/en-us/azure/storage/blobs/soft-delete-overview (Put operation)

tombuildsstuff commented:
Ran the tests for this and they pass - thanks for this @evenh / sorry for the delay getting this merged 👍

evenh commented:
Hooray, thanks :-)
